addons/website_sale/controllers/main.py

```diff
@@ -1406,76 +1406,41 @@ def extra_info(self, **post):
     # ------------------------------------------------------
 
     def _get_express_shop_payment_values(self, order, **kwargs):
-        logged_in = not request.website.is_public_user()
-        providers_sudo = request.env['payment.provider'].sudo()._get_compatible_providers(
-            order.company_id.id,
-            order.partner_id.id,
-            order.amount_total,
-            currency_id=order.currency_id.id,
-            is_express_checkout=True,
-            sale_order_id=order.id,
+        vals = sale_portal.CustomerPortal._get_payment_values(
+            self,
+            order,
+            tokens_included=None,
```

Comment:
I feel like this line is the source of several "ugly" compromises in the rest of this PR's diff.

Wouldn't it be clearer to:
1. Create a new `_get_available_tokens(**kwargs)` method.
2. Call it in `_get_payment_values` and pass the kwargs as-is.
    -> `is_express_checkout=True` is already in the kwargs
    -> this allows moving the dict key `tokens` back into the initial definition of the dict.
3. Override it in whichever module needs to change the domain of available tokens.
    -> this should greatly simplify the controllers' logic.

Reply:
Great idea! I did the suggested changes, and it does look better, thanks!